### ISSUE TITLE
MathJax Directive: find alternative for native DOM functions

### ISSUE THREAD

**Issue by tobiasschweizer** [dhlab-basel/Salsah#194]
in `src/app/view/modules/other/mathjax.directive.ts` we currently use native DOM function within the callbacks assigned to the HostListeners. This is bad practice.

Also we only check for the immediate parent and not for other ancestors.

@kilchenmann What would be the Angular way to deal with event binding in HTML that is assigned to "innerHTML"?

**Comment by musicEnfanthen** [comment]
Don't know if this is exactly what you're struggling with, but maybe it is worth to have a look at this` html-compile` component:

[https://github.com/patrikx3/angular-compile](https://github.com/patrikx3/angular-compile)

I re-used it to display clickable "internal" links on our prototype.

**Comment by tobiasschweizer** [comment]
Knora returns text values to Salsah that are actually HTML codes which are attached to the DOM (innerHTML). However, we have to bind events to these elements (this is handled by the mathjax.directive using `@HostListener`). The problem is that I did not find an alternative to using the native DOM functions for event handling.

Normally, we would use an Angular component. But in this case, the HTML comes from Knora.

Does the `angular-compile` turn external HTML into some kind of component so it can be treated as if it was a template?

**Comment by musicEnfanthen** [comment]
Yes, as I understand, this is exactly one of the use cases for angular-compile. It "post-compiles" the HTML codes that come with the returned values (what innerHTML does not, maybe by purpose?) into a dynamically compiled component. You can define a "context" (kind of "this") that should allow you to bind events.

Maybe this is somehow connected: [angular4-routerlink-inside-innerhtml-turned-to-lowercase](https://stackoverflow.com/questions/44613069/angular4-routerlink-inside-innerhtml-turned-to-lowercase)

Maybe it is also worth considering to separate the event binding (that is not only used for mathjax operations, but maybe also for showing a thumbnail preview on salsah links?) from the mathjax directive that could only be concerned with displaying the math (and with activating the event binding in any necessary case).

**Comment by tobiasschweizer** [comment]
> It "post-compiles" the HTML codes that come with the returned values

This sounds interesting. With `innerHTML`, the HTML received from Knora is just attached to the DOM, no questions asked.

> Maybe it is also worth considering to separate the event binding (that is not only used for mathjax operations, but maybe also for showing a thumbnail preview on salsah links?)

Yes, the name is misleading. Actually every text value that is sent back from Knora as HTML is passed to the directive. I called it mathjax.directive because obviously the math was the main problem. But it also adds generic event handling (standoff links).
